### ISSUE TITLE
Arch = arm64 doesn't work

### ISSUE THREAD

**Issue by tifoji** [superr/deodex_lollipop_linux#1]
I manually modified the following line 

arch="arm64" 

and removed the following line

arch=$(ls -d */ | sed 's/\///')

for a build.prop from Shield Android TV

**Comment by superr** [comment]
I am glad you got it sorted out and thanks for letting me know. I will do some testing to see what the problem is.

**Comment by tifoji** [comment]
Here is the issue
foo@bar:~/deodex_lollipop_linux/system/framework$ ls -d */ | sed 's/\///'
arm
arm64
foo@bar:~/deodex_lollipop_linux/system/framework$ 

The current command incorrectly identifies the architecture as "arm" since it is listed first, instead of "arm64" which breaks everything.

64bit android builds will be common down the road just like Shield Android TV, so it wasn't taken into account in your current script. In 64bit builds, both "arm" and "arm64" directories will be present.

These are the lines from build.prop of Shield ATV archwhich identify the arch 

ro.product.cpu.abi=arm64-v8a
ro.product.cpu.abilist64=arm64-v8a

You could detect it from build.prop as well. There are multiple solutions. I am sure you will pick up a very elegant one :) This script is amazing and helped me big time.

Apologies for not posting the above yesterday and not explain the exact scenario as I was in a rush with the builds and other work.

Thank you for your wonderful script. Great job

**Comment by superr** [comment]
Thanks for the detailed explanation. I don't have an arm64 device so I did not know both arm and arm64 directories would be present in arm64 framework, and I even assumed only one would be present. 

Please try replacing this code:

```
arch=""
cd $framedir
arch=$(ls -d */ | sed 's/\///')
```

 with this:

```
arch=""
cd $framedir
archtest=$(ls -d */ | sed 's/\///' | grep arm$)
arch64test=$(ls -d */ | sed 's/\///' | grep arm64)
archx86test=$(ls -d */ | sed 's/\///' | grep x86$)
if [[ $arch64test = "arm64" && -f $framedir/$arch64test/boot.oat ]]; then
    arch="arm64"
elif [[ $archx86test = "x86" && -f $framedir/$archx86test/boot.oat ]]; then
    arch="x86"
elif [[ $archtest = "arm" && -f $framedir/$archtest/boot.oat ]]; then
    arch="arm"
else
    while [[ ! -f $framedir/$arch/boot.oat ]]; do
        echo ""
        echo "Configure the arch of your device."
        echo ""
        echo "HINT:"
        echo "Check the framework directory, you should see"
        echo "another directory inside. The name of it should go"
        echo "here (ex. arm, arm64, x86)."
        echo ""
        echo "If you can't get past this part, check the"
        echo "following:"
        echo ""
        echo "1. Make sure this is a Lollipop rom."
        echo "2. Make sure you typed the arch variable correctly."
        echo "3. Make sure your rom is not already deodexed."
        echo ""
        echo "Type the arch of your device and press ENTER."
        echo ""
        read arch 
        clear
    done
fi
```

to see how it works for your device

Thanks for pointing this out :)

**Comment by superr** [comment]
Are there odex files in the /framework/arm directory on an arm64 device? If so, the script as it is will not deodex them if the arch is set to arm64. If you can send me a link to an odexed arm64 rom it would help me get this sorted properly. Thanks

**Comment by tifoji** [comment]
Yes there are odex files in /system/framework/arm as well as /system/framework/arm64

Actually I was planning on sending you the link for the rom. It is directly available on the nvidia site.

http://developer.download.nvidia.com/mobile/shield/ROM/SHIELD_ATV/Upgrade-1.3/extract-nv-recovery-image-shield-foster_e-32494_566.2783.zip

After extracting the zip and running the script you will get system.img and other img files.

#simg2img system.img system.img.raw   (uncompress system.img)
#sudo mount -t ext4 -o loop system.img.raw /your/mnt/point/
#cd /your/mnt/point to go /system/app/ , /system/priv-app/ and /system/framework

One more suggestion I had, since you also are accomodating x86, make it x86_64 compatiable as well to future proof it :)

**Comment by superr** [comment]
Thanks for the rom link. Not all the apps are arm64. I assume the arm apps rely on the /framework/arm odex files and the arm64 apps rely on the /framework/arm64 odex files. However, I am not sure when deodexing framework if I should pack the arm64 odex files into the corresponding framework or the arm odex files. I assume arm64.

I will put together a new script. I hope you will test it, flash the rom, and let me know how it goes :)

**Comment by superr** [comment]
Please try this out and let me know how it goes.

link removed

**Comment by superr** [comment]
Here is one with the same changes plus what I believe to be x86_64 support :)

link removed

**Comment by superr** [comment]
Found a fatal problem with both of the 2 links i provided. Try this one instead and let me know how it works out :)

link removed

**Comment by superr** [comment]
I hope you still have not downloaded any of these, otherwise you will probably be annoyed by now. I fixed a few more bugs and this one should be the last one until I hear back how it works after flashing the rom. As it is, it will deodex arm and arm64, use the relevant arm/arm64 frameworks for deodexing the app and priv-app, and use arm64 for deodexing framework. The last part is just a guess since I do not have a device to test it on. Let me know how it goes :)

https://github.com/superr/deodex_lollipop_linux/releases/download/deodex_lollipop_linux_test/deodex_lollipop_linux_test3.zip
